### PR TITLE
Add PEG parser for 'var' statements.

### PR DESCRIPTION
**Pull request by rogerl50**
This PR implements a parser, rather than regex, for dealing with dynamic expressions. The goal was to allow arbitrarily complex nested expressions, such as
`"var x = ${str1} + '.' + ${str2}"`
`"var x = sin(atan(1))"`
`"var x = 3+(4*5)/12"`
The operator priority is (hopefully) the same as in Javascript (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence#table).

As a result, I changed the syntax for the built-in functions (except for `concat`, `in`, and `includes`) from
`"var x = substr 'abc' 1 2"` or `"var x = 'abc' substr 1 2"`
to
`"var x = substr('abc',1,2)"`

In addition, the five odd functions that modified an array (`push`, `unshift`, `remove`, `insert`, `setIndex`) now use the new (and IMO more natural) syntax
`"push(a,1)"` and similar.

Finally, in all cases, in places where spaces *should* be optional, they now actually are optional, so that for example `"var x=3+4"` is valid.

This does *not* support the "abacus" symbol for replacing an operation by a variable, as I don't think it's needed any more. But certainly could be done.

The only drawback I can see to this approach is that the old syntax will be invalid and people will have to learn a new one. It has the advantage that we can improve the grammar over time to allow richer syntax (e.g. for accessing nested properties).

I also modified the dynamic testcafe test to print the offending statement if something fails, and not to print the entire log.

**Previous version of PR**
Right now, this only implements existing syntax (I hope!), with the extension that purely arithmetic expressions can be complex. For example, `(3+2)*15-4` is accepted and, one hopes, returns `71`.

I definitely need help with a couple of things:
1. The output of the PEG parser does not fit neatly into the model where all the javascript is streamed together on the client, and I can't figure out how to make that happen.
2. I'm not sure how to integrate creation of the parser output from the source (which I haven't put in the PR yet) as part of the build.
If you want to look at either of these, I can give you more details.

In addition, lots of work still remains on the grammar. For example, I'd love to be able to accept things like
`${str1} + "." + ${str2}`
`length ("abc" concat "def")`
`sin(4.5)+27`

*Update:* the code now passes all the testcafe dynamic expression tests. In addition, optional spaces are now really optional, so that `"var x=3+4"` is legal. In addition, arithmetic expressions can be arbitrarily complex, e.g. `"var x = (3+${a})*15"` is valid.

The existing syntax involving built-in functions creates some issues here. I'd like the new engine to use only function-style notation (except for `in`, `includes`, and `concat`), but it's hard to disambiguate the old-style `"var x = length 'abc'"` from `"var x = length('abc')+3"`. At present, the latter statement produces an error, although both `"var x = length 'abc'"` and `"var x = 3+length('abc')"` work properly.

I'd love some feedback on this entire idea and on the desirability of replacing the old syntax with the new one using a file updater.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3965/pr-test (or any other room on that server)

**Comment by bjalder26** [comment]
The updater seems to work if there are no functions.
image
If there ARE functions, it does not appear to work.
image

File: server/fileupdater.mjs [hunk 1]
```diff
@@ -288,3 +289,57 @@ function v5DynamicFaceProperties(properties) {
...
+  const operation  = `${identifier}`;
+
+  const regex      = `^${left} += +(?:${parameter}|(?:${parameter} +)?(🧮)?(${identifier})(?: +${parameter})?(?: +${parameter})?(?: +${parameter})?)( *|(?: +//.*))?`;
+
+  if(!Array.isArray(routine))
+    return;
+
+  for(const [i,a] of Object.entries(routine)) {
+    if (typeof a != 'string')
+      return;
```
Comment (robartsd) [inline server/fileupdater.mjs:313]: Here you use `return` which causes the function to exit. You want `continue` to skip to the next iteration of the loop instead.

Comment (bjalder26) [reply server/fileupdater.mjs:313]: Yeah, that's where it is exiting.

Comment (rogerl50) [reply server/fileupdater.mjs:313]: Thanks. I had found and fixed one of those errors, but missed this one. Still doesn't quite work (note that `{"a": "b"}` is translated as `{a: "b"}`, but at least I can move past this one.

**Comment by bjalder26** [comment]
Obviously the numbers could be replaced by variables, but is there any way the operations could be replaced by variables, or a way a string could be evaluated?  Not asking for a change per say, just wondering if that would be possible.  I think there is a workaround that could be done anyway.

**Comment by robartsd** [comment]
> Obviously the numbers could be replaced by variables, but is there any way the operations could be replaced by variables, or a way a string could be evaluated? Not asking for a change per say, just wondering if that would be possible. I think there is a workaround that could be done anyway.

Doesn't the current system have a special unicode character for using a variable for the operation. I think it was used to mitigate the compute tutorial when we swiched from COMPUTE

**Comment by bjalder26** [comment]
Maybe this is a crazy idea, but what if there was a way to remove single quotes from the var string?

For example what would normally be "var a = '2+2'" could become "var a = 2+2" and would get evaluated to 4.

As a possible syntax, maybe "var a = 🧮'2+2'" could have the 🧮 and single quotes stripped away from the string, and become "var a = 2+2".

The real benefit would be putting strings into variables or properties to be passed into dynamic routines.

For example
"a": "2+2",
"clickRoutine": [
"var a = 🧮${PROPERTY a}
]

So, if needed, game makers could concat together a string for an operation they would want to use.  It would be a different way to accomplish what using the abacus as a variable accomplished.

I would think it could be done in the JavaScript with an includes for the abacus, and then a couple of replaceAlls.

**Pull request closed without merging**